README.md
# FORKED FROM UNMAINTAINED GITHUB PROJECT

This project was originally from [gazmull/discord-paginationembed on GitHub](https://github.com/gazmull/discord-paginationembed), and was available as [discord-paginationembed](https://www.npmjs.com/package/discord-paginationembed) on NPM. However, as of December 2022, the project hasn't been updated in 1.5 years, and doesn't have support for Discord.JS v14 (the original maintainer started adding support for v13, but there was no official release for that).

Rather than bothering to update the whole library, I've updated only the parts we use to support DiscordJS v14, and removed the rest. This means that the original maintainer probably wouldn't accept a pull request for this (although to be fair, I didn't try), so we just use this fork instead.


# Discord.JS - PaginationEmbed
A pagination utility for MessageEmbed in Discord.JS

[![Discord Server](https://discordapp.com/api/guilds/370614673122263041/embed.png)](https://discord.gg/eDUzT87)
[![Travis (.org) branch](https://img.shields.io/travis/gazmull/discord-paginationembed/master.svg?logo=travis&style=flat-square)](https://travis-ci.org/gazmull/discord-paginationembed)
[![npm peer dependency version](https://img.shields.io/npm/dependency-version/discord-paginationembed/peer/discord.js.svg?logo=npm&style=flat-square)](https://nodei.co/npm/discord-paginationembed/)
[![npm type definitions](https://img.shields.io/npm/types/discord-paginationembed.svg?logo=npm&style=flat-square)](https://nodei.co/npm/discord-paginationembed/)

[![NPM](https://nodei.co/npm/discord-paginationembed.png?downloads=true&stars=true)](https://nodei.co/npm/discord-paginationembed/)

## 📣 Notice Board
- [**Changelog**](https://github.com/gazmull/discord-paginationembed/blob/master/CHANGELOG.md)
- [**Updating from `v1`**](https://github.com/gazmull/discord-paginationembed/blob/master/UPDATING_V2.md)
- Recently updated to `v2`! If **[Documentation]** feels off, please clear site cache!
- This utility no longer provide updates for v1 (Discord.JS **11**)

## 🎉 Welcome
- ✔ **Typings** included
- ✔ **[Documentation]** for online references
- ✔ **Asynchronous** workflow
- ✔ Supports [Discord.JS **12**](https://discord.js.org "Go to Discord.JS Documentation")
- ❔ Nothing found within docs or need a nudge? You may visit the [**Discord server**](https://discord.gg/eDUzT87)

## 🛠 Installation
- **Requires Discord.JS 12**: `npm install discord.js`
- **PaginationEmbed**:
    - **Published**: `npm install discord-paginationembed@beta`
    - **Unpublished**: `npm install gazmull/discord-paginationembed`
      - ❗ Requires [**Git**](https://git-scm.com/)

## 🔰 Examples
> [**Test Unit Example**](https://github.com/gazmull/discord-paginationembed/blob/master/test)

> [**Preface for TypeScript Projects**](https://github.com/gazmull/discord-paginationembed/blob/master/UPDATING_V1.md#TypeScript)

### In-action samples:
- [**Pages of command description**](https://github.com/gazmull/eros-bot/blob/master/src/commands/general/guide.ts#L35)
- [**Toggle between character and its weapon, and toggle image visibility**](https://github.com/gazmull/eros-bot/blob/master/src/commands/kamihime/info.ts#L180)
- [**EXP Leaderboard**](https://github.com/gazmull/eros-bot/blob/master/src/commands/level/leaderboard.ts#L23)
- [**Music Queue**](https://github.com/gazmull/ramiel-bot/blob/master/src/commands/music/queue.ts#L38)

<br>

### ⚠ Warning
- Examples are written under `message` event!
- There are some methods not shown in the examples. If you want to know more methods to fiddle, please visit the **[Documentation]**

### `Embeds` Mode
- A pagination mode that uses an array of MessageEmbed to paginate.
- Usually used for complex builds such as characters' information.

```js
const Discord = require('discord.js');
const Pagination = require('discord-paginationembed');

const embeds = [];

for (let i = 1; i <= 5; ++i)
  embeds.push(new Discord.MessageEmbed().addField('Page', i));

const myImage = message.author.displayAvatarURL();

new Pagination.Embeds()
  .setArray(embeds)
  .setAuthorizedUsers([message.author.id])
  .setChannel(message.channel)
  .setPageIndicator(true)
  .setPage(3)
   // Methods below are for customising all embeds
  .setImage(myImage)
  .setThumbnail(myImage)
  .setTitle('Test Title')
  .setDescription('Test Description')
  .setFooter('Test Footer Text')
  .setURL(myImage)
  .setColor(0xFF00AE)
  .addField('\u200b', '\u200b')
  .addField('Test Field 1', 'Test Field 1', true)
  .addField('Test Field 2', 'Test Field 2', true)
  .build();
```
![Embeds](https://user-images.githubusercontent.com/32944712/37118454-41116cbe-228f-11e8-9878-f39db26316a1.png)

> This assumes this is under an `async` function

```js
const Discord = require('discord.js');
const Pagination = require('discord-paginationembed');

const embeds = [];

for (let i = 1; i <= 5; ++i)
  embeds.push(new Discord.MessageEmbed().addField('Page', i));

const Embeds = new PaginationEmbed.Embeds()
  .setArray(embeds)
  .setAuthorizedUsers([message.author.id])
  .setChannel(message.channel)
  .setPageIndicator(true)
  .setTitle('Test Title')
  .setDescription('Test Description')
  .setFooter('Test Footer Text')
  .setURL('https://gazmull.github.io/discord-paginationembed')
  .setColor(0xFF00AE)
  // Sets the client's assets to utilise. Available options:
  //  - message: the client's Message object (edits the message instead of sending new one for this instance)
  //  - prompt: custom content for the message sent when prompted to jump to a page
  //      {{user}} is the placeholder for the user mention
  .setClientAssets({ message, prompt: 'Page plz {{user}}' })
  .setDeleteOnTimeout(true)
  .setDisabledNavigationEmojis(['delete'])
  .setFunctionEmojis({
    '⬆': (_, instance) => {
      for (const embed of instance.array)
        embed.fields[0].value++;
    },
    '⬇': (_, instance) => {
      for (const embed of instance.array)
        embed.fields[0].value--;
    }
  })
  // Listeners for PaginationEmbed's events
  // After the initial embed has been sent
  // (technically, after the client finished reacting with enabled navigation and function emojis).
  .on('start', () => console.log('Started!'))
  // When the instance is finished by a user reacting with `delete` navigation emoji
  // or a function emoji that throws non-Error type.
  .on('finish', (user) => console.log(`Finished! User: ${user.username}`))
  // Upon a user reacting on the instance.
  .on('react', (user, emoji) => console.log(`Reacted! User: ${user.username} | Emoji: ${emoji.name} (${emoji.id})`))
  // When the awaiting timeout is reached.
  .on('expire', () => console.warn('Expired!'))
  // Upon an occurance of error (e.g: Discord API Error).
  .on('error', console.error);

await Embeds.build();
```
![Embeds2](https://github.com/gazmull/discord-paginationembed/blob/master/demo/Embeds.gif?raw=true)

## 💡🐛💻 Contributing
### Bug Reports
Please provide reproducible steps and results proofs (e.g: images). Also, solutions are welcome!

### Suggestions / Discussions
Please be explicit about the feature's description and provide a valid reason (e.g: beneficial to users or development time) why it should be added/changed/removed.

### Source Code
- Fork this repository.
- Execute `npm install`
- Code and code and code and code and... code!
  - To enable [**incremental compilation**](https://en.wikipedia.org/wiki/Incremental_compiler) to JS: `npm run dev:watch`
- `npm test` to verify if your additions/adjustments are following the project's codebase rules and to verify if the docs are valid.
- Please make sure that you have tested your changes very well.
    - There is a test bot script under `test` folder. To get started:
        - Copy `credentials.sample.js` to `credentials.js` and fill up your private credentials (token, test channel, etc)
        - Execute either:
          - One-time test: `npm run test:bot`
          - Hot-reloading test (nodemon): `npm run dev:start`
- File a [**Pull Request (PR)**](https://github.com/gazmull/discord-paginationembed/compare)!
- For the PR comment, it goes the same with **`Suggestions / Discussions`**.

# License
> [**MIT**](https://github.com/gazmull/discord-paginationembed/blob/master/LICENSE)

© 2018-present [**Euni (gazmull)**](https://github.com/gazmull)

[Documentation]: https://gazmull.github.io/discord-paginationembed/master


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [3.0.0] - 2022-11-27
- Discord.js v14 support
- Remved `FieldEmbeds` mode

## [2.1.0] - 2020-07-22

### Added
- `Embeds` mode:
  - Added `toJSON` method.
  - `setArray` method now accepts an array of embed data (plain object):
    ```js
    <PaginationEmbed>.setArray([
        { title: 'Yes', description: 'No' },
        { title: 'No', description: 'Yes' },
    ]);
    ```

## [2.0.0] - 2020-05-27

This version involves reflection of official release of Discord.JS 12, code optimisations, and changes that will give the user more control to the instance
such as customisable page indicator, and favourable event emissions.

### Added
- (ft. [@cycloptux]) - `setPageIndicator` is now powerful for customisation.
  - First parameter has additional option `footer` (makes the indicator replace the embed's footer text instead).
    - `footer` option will not *directly* modify embeds from `array` when using `Embeds` mode.
  - Second parameter has been added for indicator's formatting function: either use the following pre-made formats `'text' | 'textcompact' | 'circle' | 'hybrid'` — or make your own format.
  - In case you don't like the placing of the indicator (i.e. you want it in embed title instead), you can set the first parameter as false but set the second parameter to your preferred format and later on use the format by accessing getter `pageIndicator` then modify the embed via `pageUpdate` event.
  - **Example**
    - ```js
      <PaginationEmbed>.setPageIndicator('footer', (page, pages) => `peij ${page} 0f ${pages} - xoxo`)
        ```
    - Pre-made formats preview:
      - **(default)** text: `Page 1 of 2`
      - textcompact: `1/2`
      - circle: `● ○`
      - hybrid: `[1/2] ● ○`
- `setAuthorizedUsers` now accepts a user ID as well instead of just an array of user IDs.

### Changed
- **⚠ BREAKING** - `Embeds` mode's changes to reflect Discord.JS' MessageEmbed methods:
  - Removed `addBlankField`
  - `spliceField` renamed to `spliceFields`
  - Added `addFields`
- **⚠ BREAKING** - Removed `I` prefix on the following TypeScript interfaces: `INavigationEmojis`, `IClientAssets`, `IFunctionEmoji`
- **⚠ BREAKING** - Renamed property `pageIndicator` to `usePageIndicator`
- **⚠ BREAKING** - Navigation emoji identifiers (`BACK`, `JUMP`, `FORWARD`, `DELETE`, `ALL`) has been lowercased due to unnecessary internal transformations.
- `usePageIndicator`'s default is now set to false.
- Event `pageUpdate` has been relocated to emit at initial page as well.
- Events' JSDoc description has been improved.
- ([@cycloptux]) - Peer dependency Discord.JS version updated to `^12.0.0`
- README's example has been updated to reduce complexity.

### Fixed
- ([@d-shaun]) - Undesirable placement of `start` event's emission

### Internal
- (ft. [@cycloptux]) - Test script `bot.js` updated for new changes
- Migrated linter from TSLint to ESLint
- Upload vscode settings

[@cycloptux]: https://github.com/cycloptux
[@d-shaun]: https://github.com/d-shaun

## [2.0.0-beta.4] - 2020-02-14
Utility has been updated to support the breaking change feature `Managers`. Starting from this version it will no longer support the [Discord.JS](https://github.com/discordjs/discord.js) master commits earlier than [`bbdbc4c`](https://github.com/discordjs/discord.js/commit/bbdbc4cfa789383b7b3dbecf5e6b8401ea2dd998).

## [2.0.0-beta.3] - 2020-01-21

### Added
- `attachFiles` method for Embeds mode
- `pageUpdate` event (emitted when the page number is updated via reaction)
- Verbose documentation to `navigationEmojis` and `functionEmojis`

### Changed
- Test script `bot.js` updated for new features

### Fixed
- Incorrect documentation examples for FieldsEmbed mode
- Fix `Unknown Message` error when `deleteOnTimeout` is enabled

## [2.0.0-beta.2] - 2019-11-12

### Fixed
- Embeds mode's `_loadList` overwriting existing message content

## [2.0.0-beta.1] - 2019-11-09

### Added
- Documentation example for IFunctionEmoji: stopping the instance from awaiting reacts with function emojis

### Changed
- Documentation changes involving `page` and `pages` properties

### Fixed
- Incorrect installation note on `UPDATING_V1.md`
- FieldsEmbed mode's `_loadList` overwriting existing message content

## [2.0.0-beta.0] - 2019-07-29
Nothing new, but incremented version to make way for updated Discord.JS v11 variant release.

## [1.0.0] - 2019-07-29
Latest release for [v11 branch](https://github.com/gazmull/discord-paginationembed/tree/v11). See changes below from 1.0.0-beta.0.

**Master branch (For Discord.JS v12)** has been incremented to `2.0.0-beta.0`

## [1.0.0-beta.4] - 2019-04-26
[**Updating from `v0.8.0`**](UPDATING_V1.md#api-changes) (Updated, please read again)

### Added
- Added ability to stop the instance from awaiting emoji reacts with function emojis:
    ```js
    <PaginationEmbed>.addFunctionEmoji('🛑', () => {
        // Either
        throw 'stopped';

        // or
        return Promise.reject('stopped');

        // will stop the instance from awaiting reacts.
        // Passing an error object will emit the `error` event.
    });
    ```
- Added `PaginationEmbed#setEmojisFunctionAfterNavigation` method. This allows function emojis to either be the first/last set to be deployed before/after navigation emojis.
  -  Default: `false`.

### Changed
- PaginationEmbed no longer emits an event when there is no listener (invoking `.on`/`.once`)

### Removed
- Option `prepare` for `clientAssets` has been removed. This affects:
    - Setting the `channel` property is a must now. (Used to be either `clientAssets.message` or `channel` must be set)

### Fixed
- Fixed possibly unnecessary API call on awaiting emoji reacts timeout.
  - `clientMessage#delete` now precedes `clientMessage.reactions#removeAll`

## [1.0.0-beta.3] - 2019-03-31
### Added
- More TypeScript notices

test/bot.js
const { AttachmentBuilder, Client, GatewayIntentBits } = require('discord.js');
const PaginationEmbed = require('../');

const credentials = require('./credentials');

const bot = new Client({
  intents: [
    GatewayIntentBits.MessageContent,
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.GuildMessageReactions,
    GatewayIntentBits.DirectMessages,
    GatewayIntentBits.DirectMessageReactions,
  ],
});

const error = msg => {
  console.error(msg);

  process.exit(1);
};

const done = () => {
  console.log('Test done!');

  process.exit(0);
};

bot
  .on('ready', async () => {
    const channel = bot.channels.cache.get(credentials.channel);

    if (!channel)
      error('Provided channel is not resolvable by the client.');

    console.log('Ready to test! Channel name:', channel.name, 'Channel ID:', channel.id);

    const {
      test,
      users,
      disabledNavigationEmojis,
      emojisFunctionAfterNavigation,
      deleteOnTimeout
    } = credentials;

    console.log('Mode:', test);

    if (true) {
      const embeds = [];

      for (let i = 1; i <= 3; ++i)
        embeds.push({
          fields: [ { name: 'Page', value: i, inline: false } ],
          image: { url: 'attachment://1.jpg' }
        });

      const Embeds = new PaginationEmbed.Embeds()
        .setArray(embeds)
        .setFile(new AttachmentBuilder(`${__dirname}/images/1.jpg`))
        .setAuthorizedUsers(users)
        .setChannel(channel)
        .setTitle('Test Title')
        .setDescription('Test Description')
        .setPageIndicator(true, 'hybrid')
        .setContent('benis', '\n\n')
        .setFooter(`version: ${PaginationEmbed.version}`)
        .setURL('https://gazmull.github.io/discord-paginationembed')
        .setColor(0xFF00AE)
        .addField('Test Field 1', 'Test Field 1', true)
        .addField('Test Field 2', 'Test Field 2', true)
        .setDeleteOnTimeout(deleteOnTimeout)
        .setEmojisFunctionAfterNavigation(emojisFunctionAfterNavigation)
        .setDisabledNavigationEmojis(disabledNavigationEmojis)
        .setFunctionEmojis({
          '⬆': (_, instance) => {
            for (const embed of instance.array)
              embed.data.fields[0].value++;
          },
          '⬇': (_, instance) => {
            for (const embed of instance.array)
              embed.data.fields[0].value--;
          },
          '⏹': () => Promise.reject('stopped'),
          '🔕': () => Promise.reject(new Error('Worst Error Ever'))
        })
        .setClientAssets({ prompt: 'yAAAaA— what page {{user}}?' })
        .on('start', () => console.log('Started!'))
        .on('finish', (user) => console.log(`Finished! User: ${user.username}`))
        .on('react', (user, emoji) =>
          console.log(`Reacted! User: ${user.username} | Emoji: ${emoji.name} (${emoji.id})`))
        .on('pageUpdate', () => Embeds.currentEmbed.title = Embeds.pageIndicator)
        .on('expire', () => console.warn('Expired!'))
        .on('error', console.error);

      await Embeds.build();

      return done();
    }
  })
  .login(credentials.token);


test/credentials.sample.js
module.exports = {
  // Your bot's token
  token: null,
  // The channel where to send the embed test
  channel: null,
  // Users authorized to interact with the utility
  users: [],
  // 'all' / 'back' / 'jump' / 'foward' / 'delete',
  disabledNavigationEmojis: [ 'delete' ],
  // Whether function emojis should be deployed after navigation emojis
  emojisFunctionAfterNavigation: false,
  // Delete PaginationEmbed message after awaiting response timeout?
  deleteOnTimeout: true
};


typings/Embeds.d.ts
import { APIEmbed, AttachmentBuilder, ColorResolvable, EmbedBuilder } from 'discord.js';
import { PaginationEmbed } from './base';
/**
 * A pagination mode that uses an array of EmbedBuilder to paginate.
 * @extends [[PaginationEmbed]]
 * @noInheritDoc
 */
export declare class Embeds extends PaginationEmbed<EmbedBuilder> {
    /** The title of all embeds. */
    title: string;
    /** The description of all embeds. */
    description: string;
    /** The URL of all embeds. */
    url: string;
    /** The color of all embeds. */
    color: number;
    /** The timestamp of all embeds. */
    timestamp: number;
    /** The thumbnail of all embeds. */
    thumbnail: string;
    /** The image of all embeds. */
    image: string;
    /** The author of all embeds. */
    author: {
        name: string;
        url?: string;
        iconURL?: string;
    };
    /** The footer of all embeds. */
    footer: {
        text: string;
        iconURL?: string;
    };
    /** Embed in the current page. */
    get currentEmbed(): EmbedBuilder;
    get currentFiles(): AttachmentBuilder[];
    get pages(): number;
    /**
     * Adds a field to the fields of all embeds.
     * @param name - The name of the field.
     * @param value - The value of the field.
     * @param inline - Whether the field is inline to the other fields.
     */
    addField(name: string, value: string, inline?: boolean): this;
    /**
     * Build the Pagination Embeds.
     *
     * ### Example
     * ```js
     *   const { Embeds } = require('discord-paginationembed');
     *   const { EmbedBuilder } = require('discord.js');
     *
     *   // Under message event.
     *   const embeds = [];
     *
     *   for (let i = 0; i < 5; ++i)
     *    embeds.push(new EmbedBuilder().addFields({name: 'Page', value: i + 1}));
     *
     *   new Embeds()
     *    .setAuthorizedUsers([message.author.id])
     *    .setChannel(message.channel)
     *    .setClientAssets({ prompt: 'Yo {{user}} wat peige?!?!?' })
     *    .setArray(embeds)
     *    .setPageIndicator(false)
     *    .setPage(1)
     *    .setTimeout(69000)
     *    .setNavigationEmojis({
     *      back: '◀',
     *      jump: '↗',
     *      forward: '▶',
     *      delete: '🗑'
     *    })
     *    .setFunctionEmojis({
     *      '⬆': (_, instance) => {
     *        for (const embed of instance.array)
     *          embed.fields[0].value++;
     *        },
     *      '⬇': (_, instance) => {
     *         for (const embed of instance.array)
     *           embed.fields[0].value--;
     *        }
     *    })
     *    .setDescription('This is one of my embeds with this message!')
     *    .setColor(0xFF00AE)
     *    .setTimestamp()
     *    .build();```
     */
    build(): Promise<void>;
    /**
     * Sets the array of EmbedBuilder to paginate.
     * @param array - An array of EmbedBuilder to paginate.
     */
    setArray(array: (APIEmbed | EmbedBuilder)[]): this;
    /**
     * Set the author of all embeds.
     * @param name - The name of the author.
     * @param iconURL - The icon URL of the author.
     * @param url - The URL of the author.
     */
    setAuthor(name: string, iconURL?: string, url?: string): this;
    /**
     * Sets the color of all embeds.
     * @param color - The color of all embeds.
     */
    setColor(color: ColorResolvable): this;
    /**
     * Sets the description of all embeds.
     * @param description - The description of all embeds.
     */
    setDescription(description: string): this;
    /**
     * Sets the footer of all embeds.
     * @param text - The footer text.
     * @param iconURL - URL for the footer's icon.
     */
    setFooter(text: string, iconURL?: string): this;
    /**
     * Sets the image of all embeds.
     * @param url - The image of all embeds.
     */
    setImage(url: string): this;
    /**
     * Sets the thumbnail of all embeds.
     * @param url - The thumbnail of all embeds.
     */
    setThumbnail(url: string): this;
    /**
     * Sets the timestamp of all embeds.
     * @param timestamp - The timestamp or date.
     */
    setTimestamp(timestamp?: Date | number): this;
    /**
     * Sets the title of all embeds.
     * @param title - The title of all embeds.
     */
    setTitle(title: string): this;
    /**
     * Sets the URL of all embeds.
     * @param url - The URL of all embeds.
     */
    setURL(url: string): this;
    /**
     * Removes, replaces, and inserts fields in all embeds (max 25).
     * @param index - The index to start at.
     * @param deleteCount - The number of fields to remove.
     * @param name - The name of the field.
     * @param value - The value of the field.
     * @param inline - Set the field to display inline.
     */
    spliceFields(index: number, deleteCount: number, name?: string, value?: string, inline?: boolean): this;
    /** Transforms all embeds to plain objects. */
    toJSON(): APIEmbed[];
    /** @ignore */
    _loadList(callNavigation?: boolean): Promise<void>;
}


src/Embeds.ts
import {
  APIEmbed,
  AttachmentBuilder,
  ColorResolvable,
  EmbedBuilder,
  Message
} from 'discord.js';
import { PaginationEmbed } from './base';

/**
 * A pagination mode that uses an array of EmbedBuilder to paginate.
 * @extends [[PaginationEmbed]]
 * @noInheritDoc
 */
export class Embeds extends PaginationEmbed<EmbedBuilder> {

  /** The title of all embeds. */
  public title: string;

  /** The description of all embeds. */
  public description: string;

  /** The URL of all embeds. */
  public url: string;

  /** The color of all embeds. */
  public color: number;

  /** The timestamp of all embeds. */
  public timestamp: number;

  /** The thumbnail of all embeds. */
  public thumbnail: string;

  /** The image of all embeds. */
  public image: string;

  /** The author of all embeds. */
  public author: {
    name: string;
    url?: string;
    iconURL?: string;
  };

  /** The footer of all embeds. */
  public footer: {
    text: string;
    iconURL?: string;
  };

  /** Embed in the current page. */
  get currentEmbed (): EmbedBuilder {
    return this.array[this.page - 1];
  }

  /* Files to attach to the current page. */
  get currentFiles (): AttachmentBuilder[] {
    return this.files ? [ this.files[this.page - 1] ] : [];
  }

  get pages (): number {
    return this.array.length;
  }

  /**
   * Adds a field to the fields of all embeds.
   * @param name - The name of the field.
   * @param value - The value of the field.
   * @param inline - Whether the field is inline to the other fields.
   */
  public addField (name: string, value: string, inline = false) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.addFields({name, value, inline});

    return this;
  }

  /**
   * Build the Pagination Embeds.
   *
   * ### Example
   * ```js
   *   const { Embeds } = require('discord-paginationembed');
   *   const { EmbedBuilder } = require('discord.js');
   *
   *   // Under message event.
   *   const embeds = [];
   *
   *   for (let i = 0; i < 5; ++i)
   *    embeds.push(new EmbedBuilder().addFields({name: 'Page', value: i + 1}));
   *
   *   new Embeds()
   *    .setAuthorizedUsers([message.author.id])
   *    .setChannel(message.channel)
   *    .setClientAssets({ prompt: 'Yo {{user}} wat peige?!?!?' })
   *    .setArray(embeds)
   *    .setPageIndicator(false)
   *    .setPage(1)
   *    .setTimeout(69000)
   *    .setNavigationEmojis({
   *      back: '◀',
   *      jump: '↗',
   *      forward: '▶',
   *      delete: '🗑'
   *    })
   *    .setFunctionEmojis({
   *      '⬆': (_, instance) => {
   *        for (const embed of instance.array)
   *          embed.fields[0].value++;
   *        },
   *      '⬇': (_, instance) => {
   *         for (const embed of instance.array)
   *           embed.fields[0].value--;
   *        }
   *    })
   *    .setDescription('This is one of my embeds with this message!')
   *    .setColor(0xFF00AE)
   *    .setTimestamp()
   *    .build();```
   */
  public async build () {
    await this._verify();

    return this._loadList();
  }

  /**
   * Sets the array of EmbedBuilder to paginate.
   * @param array - An array of EmbedBuilder to paginate.
   */
  public setArray (array: (APIEmbed|EmbedBuilder)[]) {
    const isValidArray = Array.isArray(array) && Boolean(array.length);

    if (!isValidArray) throw new TypeError('Cannot invoke Embeds class without a valid array to paginate.');

    for (const [ i, v ] of array.entries())
      if (Boolean(v) && v.constructor === Object && Object.keys(v).length)
        array[i] = new EmbedBuilder(v as APIEmbed);
      else if (v instanceof EmbedBuilder)
        continue;
      else
        throw new TypeError(`(EmbedBuilders[${i}]) Cannot invoke Embeds class with an invalid EmbedBuilder instance.`);

    this.array = array as EmbedBuilder[];

    return this;
  }

  /**
   * Set the author of all embeds.
   * @param name - The name of the author.
   * @param iconURL - The icon URL of the author.
   * @param url - The URL of the author.
   */
  public setAuthor (name: string, iconURL?: string, url?: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setAuthor({name, iconURL, url});

    return this;
  }

  /**
   * Sets the color of all embeds.
   * @param color - The color of all embeds.
   */
  public setColor (color: ColorResolvable) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setColor(color);

    return this;
  }

  /**
   * Sets the description of all embeds.
   * @param description - The description of all embeds.
   */
  public setDescription (description: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setDescription(description);

    return this;
  }

  /**
   * Sets the footer of all embeds.
   * @param text - The footer text.
   * @param iconURL - URL for the footer's icon.
   */
  public setFooter (text: string, iconURL?: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setFooter({text, iconURL});

    return this;
  }

  /**
   * Sets the image of all embeds.
   * @param url - The image of all embeds.
   */
  public setImage (url: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setImage(url);

    return this;
  }

  /**
   * Sets the thumbnail of all embeds.
   * @param url - The thumbnail of all embeds.
   */
  public setThumbnail (url: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setThumbnail(url);

    return this;
  }

  /**
   * Sets the timestamp of all embeds.
   * @param timestamp - The timestamp or date.
   */
  public setTimestamp (timestamp?: Date | number) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setTimestamp(timestamp);

    return this;
  }

  /**
   * Sets the title of all embeds.
   * @param title - The title of all embeds.
   */
  public setTitle (title: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setTitle(title);

    return this;
  }

  /**
   * Sets the URL of all embeds.
   * @param url - The URL of all embeds.
   */
  public setURL (url: string) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.setURL(url);

    return this;
  }

  /**
   * Removes, replaces, and inserts fields in all embeds (max 25).
   * @param index - The index to start at.
   * @param deleteCount - The number of fields to remove.
   * @param name - The name of the field.
   * @param value - The value of the field.
   * @param inline - Set the field to display inline.
   */
  public spliceFields (
    index: number,
    deleteCount: number,
    name?: string,
    value?: string,
    inline?: boolean
  ) {
    if (!this.array) throw new TypeError('this.array must be set first.');

    for (const el of this.array)
      el.spliceFields(index, deleteCount, { name, value, inline });

    return this;
  }

  /** Transforms all embeds to plain objects. */
  public toJSON () {
    if (!this.array) throw new TypeError('this.array must be set first.');

    return this.array.map(m => m.toJSON());
  }

  /** @ignore */
  public async _loadList (callNavigation = true) {
    if (this.listenerCount('pageUpdate')) this.emit('pageUpdate');

    const embed = new EmbedBuilder(this.currentEmbed.data);
    const files = this.currentFiles;
    const isFooter = this.usePageIndicator === 'footer';
    const shouldIndicate = this.usePageIndicator && !isFooter
      ? this.pages === 1
        ? ''
        : this.pageIndicator
      : '';
    const { separator, text } = this.content;
    // Fixes no-argument TS error
    const content = `${text ? `${text}${separator}` : ''}${shouldIndicate}`;
    const options = { embeds: [ embed ], content: content || null, files };

    if (isFooter)
      embed.setFooter({text: this.pageIndicator, iconURL: embed.data.footer.icon_url});
    if (this.clientAssets.message)
      await this.clientAssets.message.edit(options);
    else
      this.clientAssets.message = await this.channel.send(options) as Message;

    return super._loadList(callNavigation);
  }
}
